File: scripts/plans.py
"""
Single source of truth for paid plans + quota enforcement (Python side).

Mirror of webapp/functions/_shared/plans.js — keep them byte-compatible.
Quotas are counted in `usage_counters(user_id, metric, period_key, count)`.
period_key is computed in the USER'S timezone so limits are fair globally:
  - daily metrics   → "2026-07-10"
  - weekly metrics  → "2026-W28"  (ISO year-week)
"""

from __future__ import annotations

import datetime as _dt

from d1 import execute, query

UNLIMITED = 999  # practical ceiling — protects LLM cost, never shown as a wall

# Metrics: interview, autoapply, notif, cvprep, countries
PLANS: dict[str, dict] = {
    "free":    {"interview": 1, "autoapply": 1,  "notif": 3,         "cvprep": 1,         "countries": 1},
    "starter": {"interview": 1, "autoapply": 5,  "notif": 10,        "cvprep": 5,         "countries": 3},
    "pro":     {"interview": 3, "autoapply": 15, "notif": 25,        "cvprep": 15,        "countries": UNLIMITED},
    "proplus": {"interview": UNLIMITED, "autoapply": 40, "notif": UNLIMITED, "cvprep": UNLIMITED, "countries": UNLIMITED},
}

# Display metadata (used by the dashboard / Phase 2). Prices in USD / month.
PLAN_META: dict[str, dict] = {
    "free":    {"label": "Free",     "price": 0},
    "starter": {"label": "Starter",  "price": 5},
    "pro":     {"label": "Pro",      "price": 12},
    "proplus": {"label": "Pro Plus", "price": 25},
}


def _plan(plan: str) -> dict:
    return PLANS.get((plan or "free").lower(), PLANS["free"])


def metric_limit(plan: str, metric: str) -> tuple[int, str]:
    """Return (limit, period) for a plan+metric. Free's interview resets weekly."""
    limit = int(_plan(plan).get(metric, 0))
    period = "week" if (metric == "interview" and (plan or "free").lower() == "free") else "day"
    return limit, period


def period_key(period: str, tz: str = "UTC") -> str:
    """Current period key in the user's timezone."""
    try:
        from zoneinfo import ZoneInfo
        now = _dt.datetime.now(ZoneInfo(tz or "UTC"))
    except Exception:
        now = _dt.datetime.now(_dt.timezone.utc)
    if period == "week":
        iso = now.isocalendar()
        return f"{iso[0]}-W{iso[1]:02d}"
    return now.strftime("%Y-%m-%d")


def usage_count(user_id: str, metric: str, pkey: str) -> int:
    rows = query(
        "SELECT count FROM usage_counters WHERE user_id=? AND metric=? AND period_key=?",
        [user_id, metric, pkey])
    return int(rows[0]["count"] or 0) if rows else 0


def bump(user_id: str, metric: str, pkey: str, n: int = 1) -> None:
    execute(
        "INSERT INTO usage_counters (user_id, metric, period_key, count) VALUES (?,?,?,?) "
        "ON CONFLICT(user_id, metric, period_key) DO UPDATE SET count = count + excluded.count",
        [user_id, metric, pkey, n])


def remaining(user_id: str, plan: str, metric: str, tz: str = "UTC") -> int:
    limit, period = metric_limit(plan, metric)
    used = usage_count(user_id, metric, period_key(period, tz))
    return max(0, limit - used)


def allow(user_id: str, plan: str, metric: str, tz: str = "UTC") -> bool:
    """True if the user is still under their limit (does NOT consume)."""
    limit, period = metric_limit(plan, metric)
    return usage_count(user_id, metric, period_key(period, tz)) < limit


def consume(user_id: str, plan: str, metric: str, tz: str = "UTC") -> bool:
    """Consume one credit. True if allowed (and counted), False if over limit."""
    limit, period = metric_limit(plan, metric)
    pkey = period_key(period, tz)
    if usage_count(user_id, metric, pkey) >= limit:
        return False
    bump(user_id, metric, pkey)
    return True


def once_per_day(user_id: str, tag: str, tz: str = "UTC") -> bool:
    """First call today (user tz) returns True, then False — for one-shot nudges."""
    pkey = period_key("day", tz)
    metric = f"nudge:{tag}"
    if usage_count(user_id, metric, pkey) >= 1:
        return False
    bump(user_id, metric, pkey)
    return True

File: webapp/functions/_shared/plans.js
// Single source of truth for paid plans + quota enforcement (Functions side).
// Mirror of scripts/plans.py — keep them byte-compatible.
// Quotas live in usage_counters(user_id, metric, period_key, count).
// period_key is computed in the USER'S timezone (daily "2026-07-10" or ISO
// weekly "2026-W28") so limits are fair globally.

export const UNLIMITED = 999; // practical ceiling — protects LLM cost, never shown as a wall

export const PLANS = {
  free:    { interview: 1, autoapply: 1,  notif: 3,         cvprep: 1,         countries: 1 },
  starter: { interview: 1, autoapply: 5,  notif: 10,        cvprep: 5,         countries: 3 },
  pro:     { interview: 3, autoapply: 15, notif: 25,        cvprep: 15,        countries: UNLIMITED },
  proplus: { interview: UNLIMITED, autoapply: 40, notif: UNLIMITED, cvprep: UNLIMITED, countries: UNLIMITED },
};

export const PLAN_META = {
  free:    { label: "Free",     price: 0 },
  starter: { label: "Starter",  price: 5 },
  pro:     { label: "Pro",      price: 12 },
  proplus: { label: "Pro Plus", price: 25 },
};

function planOf(plan) { return PLANS[(plan || "free").toLowerCase()] || PLANS.free; }

export function metricLimit(plan, metric) {
  const limit = Number(planOf(plan)[metric] || 0);
  const period = (metric === "interview" && (plan || "free").toLowerCase() === "free") ? "week" : "day";
  return { limit, period };
}

// --- period keys (must match plans.py exactly) --------------------------- //
function tzParts(tz) {
  // Returns {y, m, d} for "now" in the given IANA timezone.
  const fmt = new Intl.DateTimeFormat("en-CA", {
    timeZone: tz || "UTC", year: "numeric", month: "2-digit", day: "2-digit",
  });
  const parts = {};
  for (const p of fmt.formatToParts(new Date())) if (p.type !== "literal") parts[p.type] = p.value;
  return { y: +parts.year, m: +parts.month, d: +parts.day };
}

function isoWeekKey(y, m, d) {
  const dt = new Date(Date.UTC(y, m - 1, d));
  const day = (dt.getUTCDay() + 6) % 7;        // Mon=0..Sun=6
  dt.setUTCDate(dt.getUTCDate() - day + 3);     // Thursday of this ISO week
  const firstThu = new Date(Date.UTC(dt.getUTCFullYear(), 0, 4));
  const week = 1 + Math.round(((dt - firstThu) / 86400000 - 3 + ((firstThu.getUTCDay() + 6) % 7)) / 7);
  return `${dt.getUTCFullYear()}-W${String(week).padStart(2, "0")}`;
}

export function periodKey(period, tz = "UTC") {
  const { y, m, d } = tzParts(tz);
  if (period === "week") return isoWeekKey(y, m, d);
  return `${y}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}

// --- counters ------------------------------------------------------------- //
export async function usageCount(env, userId, metric, pkey) {
  const r = await env.DB.prepare(
    "SELECT count FROM usage_counters WHERE user_id=? AND metric=? AND period_key=?"
  ).bind(userId, metric, pkey).first();
  return r ? Number(r.count || 0) : 0;
}

export async function bump(env, userId, metric, pkey, n = 1) {
  await env.DB.prepare(
    "INSERT INTO usage_counters (user_id, metric, period_key, count) VALUES (?,?,?,?) " +
    "ON CONFLICT(user_id, metric, period_key) DO UPDATE SET count = count + excluded.count"
  ).bind(userId, metric, pkey, n).run();
}

export async function remaining(env, userId, plan, metric, tz = "UTC") {
  const { limit, period } = metricLimit(plan, metric);
  const used = await usageCount(env, userId, metric, periodKey(period, tz));
  return Math.max(0, limit - used);
}

// Consume one credit. Returns true if allowed (and counted), false if over limit.
export async function consume(env, userId, plan, metric, tz = "UTC") {
  const { limit, period } = metricLimit(plan, metric);
  const pkey = periodKey(period, tz);
  const used = await usageCount(env, userId, metric, pkey);
  if (used >= limit) return false;
  await bump(env, userId, metric, pkey);
  return true;
}

// Per-metric usage snapshot for the dashboard meters.
export async function usageSummary(env, userId, plan, tz = "UTC") {
  const metrics = ["notif", "autoapply", "cvprep", "interview"];
  const out = {};
  for (const m of metrics) {
    const { limit, period } = metricLimit(plan, m);
    const used = await usageCount(env, userId, m, periodKey(period, tz));
    out[m] = { used, limit, remaining: Math.max(0, limit - used), period, unlimited: limit >= UNLIMITED };
  }
  return out;
}

// The user's IANA timezone from their saved settings (falls back to UTC).
export async function userTimezone(env, userId) {
  try {
    const r = await env.DB.prepare("SELECT settings FROM configs WHERE user_id=?").bind(userId).first();
    const s = r?.settings ? JSON.parse(r.settings) : {};
    return s.timezone || "UTC";
  } catch { return "UTC"; }
}
